src/modules/job-manager/alloc.c

```diff
@@ -175,6 +175,51 @@ int cancel_request (struct alloc *alloc, struct job *job)
     return 0;
 }
 
+/* we want to delete items set to 'null', so this is not the same
+ * as json_object_update_recursive() in jansson 2.13.1
+ */
+static void update_recursive (flux_t *h, struct job *job,
+                              json_t *orig, json_t *new)
```

Comment:
This function seems like it should have unit test coverage, since it's rather general and standalone.
In fact would it belong in libutil as a reusable thing?

Maybe the error reporting could be via a passed in function plus void *arg to avoid the flux_t entanglement?

Reply:
Just a question but are we allowing any JSON type as the value of an annotation, at any nesting level? 

(Can't quite decide if that's a good idea or not.  Will it complicate `flux jobs` or other tools parsing/sorting on the annotations?)

Reply:
i actually considered adding a unit test, but the `flux_t` and such was annoying.  I like your idea of passing in a logging function.  Lets go with that.

Reply:
> Just a question but are we allowing any JSON type as the value of an annotation, at any nesting level?

There was a discussion about this earlier and we decided on any JSON type.  Although I guess we didn't explicitly say this in RFC27.

For `flux-jobs` I think it works out far easier.  B/c schedulers/users can come up with any annotations they want, it's convenient to allow a notation like "annotations.user.foo.bar" to be allowed, and each period means a sub-object.  In contrast to allowing "sched.resource_summary" as a key at the top level.